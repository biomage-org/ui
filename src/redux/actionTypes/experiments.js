const EXPERIMENTS = 'experiments';

/**
 * Create experiments
 */
const EXPERIMENTS_CREATED = `${EXPERIMENTS}/created`;

/**
 * Create experiments
 */
const EXPERIMENTS_LOADING = `${EXPERIMENTS}/loading`;

/**
 * Create experiments
 */
const EXPERIMENTS_LOADED = `${EXPERIMENTS}/loaded`;

/**
 * Experiments updated
 */
const EXPERIMENTS_UPDATED = `${EXPERIMENTS}/updated`;

/**
 * Delete experiments
 */
const EXPERIMENTS_DELETED = `${EXPERIMENTS}/deleted`;

/**
 * Saving experiment
 */
const EXPERIMENTS_SAVING = `${EXPERIMENTS}/saving`;

/**
 * Update experiments
 */
const EXPERIMENTS_ERROR = `${EXPERIMENTS}/error`;

const EXPERIMENTS_SWITCH = `${EXPERIMENTS}/switch`;

/**
 * Set active project.
 */
const PROJECTS_SET_ACTIVE = `${EXPERIMENTS}/setActive`;

/**
 * Add project metadata.
 */
const PROJECTS_METADATA_CREATE = `${EXPERIMENTS}/metadataCreate`;

/**
 * Update project metadata.
 */
const PROJECTS_METADATA_UPDATE = `${EXPERIMENTS}/metadataUpdate`;

/**
 * Delete project metadta.
 */
const PROJECTS_METADATA_DELETE = `${EXPERIMENTS}/metadataDelete`;

export {
  EXPERIMENTS_CREATED,
  EXPERIMENTS_UPDATED,
  EXPERIMENTS_ERROR,
  EXPERIMENTS_LOADING,
  EXPERIMENTS_LOADED,
  EXPERIMENTS_DELETED,
  EXPERIMENTS_SAVING,
  EXPERIMENTS_SWITCH,
  PROJECTS_SET_ACTIVE,
  PROJECTS_METADATA_CREATE,
  PROJECTS_METADATA_UPDATE,
  PROJECTS_METADATA_DELETE,
};
