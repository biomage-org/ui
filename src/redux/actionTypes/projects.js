const PROJECTS = 'projects';

/**
 * Update project.
 */
const PROJECTS_UPDATE = `${PROJECTS}/update`;

/**
 * Set active project.
 */
const PROJECTS_SET_ACTIVE = `${PROJECTS}/setActive`;

/**
 * Saving project.
 */
const PROJECTS_SAVING = `${PROJECTS}/saving`;

/**
 * Project saved.
 */
const PROJECTS_SAVED = `${PROJECTS}/saved`;

/**
 * Error saving or updating project.
 */
const PROJECTS_ERROR = `${PROJECTS}/error`;

/**
 * Add project metadata.
 */
const PROJECTS_METADATA_CREATE = `${PROJECTS}/metadataCreate`;

/**
 * Update project metadata.
 */
const PROJECTS_METADATA_UPDATE = `${PROJECTS}/metadataUpdate`;

/**
 * Delete project metadta.
 */
const PROJECTS_METADATA_DELETE = `${PROJECTS}/metadataDelete`;

/**
 * Load the projects
 */
const PROJECTS_LOADED = `${PROJECTS}/loaded`;

const PROJECTS_LOADING = `${PROJECTS}/loading`;
export {
  PROJECTS_UPDATE,
  PROJECTS_SET_ACTIVE,
  PROJECTS_ERROR,
  PROJECTS_SAVED,
  PROJECTS_SAVING,
  PROJECTS_METADATA_CREATE,
  PROJECTS_METADATA_UPDATE,
  PROJECTS_METADATA_DELETE,
  PROJECTS_LOADED,
  PROJECTS_LOADING,
};
