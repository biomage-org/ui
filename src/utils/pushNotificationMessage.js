import { message } from 'antd';

const pushNotificationMessage = (type, text) => {
  switch (type) {
    case 'success':
      message.success(text, 2);
      break;
    case 'error':
      message.error(text, 4);
      break;
    case 'info':
      message.info(text, 4);
      break;
    case 'warning':
    case 'warn':
      message.warn(text, 4);
      break;
    case 'loading':
      message.loading(text, 2);
      break;
    default:
      message.info(text, 4);
      break;
  }
};

export default pushNotificationMessage;
