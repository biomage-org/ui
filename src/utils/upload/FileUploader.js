/* eslint-disable no-unused-expressions */
import _ from 'lodash';

import { AsyncGzip } from 'fflate';
import filereaderStream from 'filereader-stream';

import fetchAPI from 'utils/http/fetchAPI';
import putInS3 from 'utils/upload/putInS3';
import UploadStatus from 'utils/upload/UploadStatus';

const fiveMB = 5 * 1024 * 1024;

class FileUploader {
  constructor(
    file,
    compress,
    chunkSize,
    uploadParams,
    abortController,
    onStatusUpdate,
  ) {
    if (!file
      || !chunkSize
      || !uploadParams
      || !abortController
      || !onStatusUpdate
    ) {
      throw new Error('FileUploader: Missing required parameters');
    }

    this.file = file;
    this.compress = compress;
    this.chunkSize = chunkSize;
    this.uploadParams = uploadParams;

    // Upload related callbacks and handling
    this.onStatusUpdate = onStatusUpdate;
    this.abortController = abortController;

    // Stream handling
    this.totalChunks = Math.ceil(file.size / chunkSize);
    this.pendingChunks = this.totalChunks;

    // Used to assign partNumbers to each chunk
    this.partNumberIt = 0;

    this.readStream = null;
    this.gzipStream = null;

    this.uploadedParts = [];

    this.resolve = null;
    this.reject = null;

    // To track upload progress
    this.uploadedPartPercentages = new Array(this.totalChunks).fill(0);

    this.currentChunk = null;

    this.accumulatedChunks = [];
  }

  async upload() {
    return new Promise((resolve, reject) => {
      this.resolve = resolve;
      this.reject = reject;

      this.readStream = filereaderStream(
        this.file?.fileObject || this.file, { chunkSize: this.chunkSize },
      );

      this.#setupReadStreamHandlers();

      if (this.compress) {
        this.gzipStream = new AsyncGzip({ level: 1, consume: true });
        this.#setupGzipStreamHandlers();
      }
    });
  }

  #uploadChunk = async (chunk) => {
    this.accumulatedChunks.push(chunk);

    const uploadSize = _.sum(_.map(this.accumulatedChunks, 'length'));
    // Upload if we have accumulated 5MB of size or if it's the last chunk (where this restriction doesn't exist)
    const canUpload = uploadSize > fiveMB || this.pendingChunks === 1;

    if (!canUpload) return;

    this.partNumberIt += 1;
    const partNumber = this.partNumberIt;

    const signedUrl = await this.#getSignedUrlForPart(partNumber);

    const mergedChunks = new Uint8Array(uploadSize);
    this.accumulatedChunks.reduce((offset, chunk) => {
      mergedChunks.set(chunk, offset);

      return offset + chunk.length;
    }, 0);

    const partResponse = await putInS3(
      mergedChunks,
      signedUrl,
      this.abortController,
      this.#createOnUploadProgress(partNumber),
    );

    this.accumulatedChunks = [];
    this.uploadedParts.push({ ETag: partResponse.headers.etag, PartNumber: partNumber });
  }

  #getSignedUrlForPart = async (partNumber) => {
    const {
      experimentId, uploadId, bucket, key,
    } = this.uploadParams;

    const queryParams = new URLSearchParams({ bucket, key });
    const url = `/v2/experiments/${experimentId}/upload/${uploadId}/part/${partNumber}/signedUrl?${queryParams}`;

    return await fetchAPI(url, { method: 'GET' });
  };

  #createOnUploadProgress = (partNumber) => (progress) => {
    // partNumbers are 1-indexed, so we need to subtract 1 for the array index
    this.uploadedPartPercentages[partNumber - 1] = progress.progress;

    const percentage = _.mean(this.uploadedPartPercentages) * 100;
    this.onStatusUpdate(UploadStatus.UPLOADING, Math.floor(percentage));
  };

  #setupGzipStreamHandlers = () => {
    this.gzipStream.ondata = async (err, chunk) => {
      try {
        if (err) throw new Error(err);

        await this.#handleChunkLoadFinished(chunk);
      } catch (e) {
        this.#cancelExecution(UploadStatus.FILE_READ_ERROR, e);
      }
    };
  }

  #setupReadStreamHandlers = () => {
    this.readStream.on('data', async (chunk) => {
      try {
        if (!this.compress) {
          // If not compressing, the load finishes as soon as the chunk is read
          await this.#handleChunkLoadFinished(chunk);
          return;
        }

        // This is necessary to connect the streams between read and compress.
        // gzipStream needs to know which is the last chunk when it is being pushed
        // but normal streams don't have a way of knowing that in the 'data' event
        // So we need to push the last chunk in the 'end' event.
        if (this.currentChunk !== null) this.gzipStream.push(this.currentChunk);

        this.currentChunk = chunk;
      } catch (e) {
        this.#cancelExecution(UploadStatus.FILE_READ_ERROR, e);
      }
    });

    this.readStream.on('error', (e) => {
      this.#cancelExecution(UploadStatus.FILE_READ_ERROR, e);
    });

    this.readStream.on('end', () => {
      try {
        if (!this.compress) return;

        this.gzipStream.push(this.currentChunk, true);
      } catch (e) {
        this.#cancelExecution(UploadStatus.FILE_READ_ERROR, e);
      }
    });
  }

  #cancelExecution = (status, e) => {
    this.readStream.destroy();

    this.gzipStream?.terminate();
    this.abortController?.abort();

    this.onStatusUpdate(status);

    this.reject(e);
    console.error(e);
  }

  #handleChunkLoadFinished = async (chunk) => {
    try {
      await this.#uploadChunk(chunk);
    } catch (e) {
      this.#cancelExecution(UploadStatus.UPLOAD_ERROR, e);
    }

    // To track when all chunks have been uploaded
    this.pendingChunks -= 1;

    if (this.pendingChunks === 0) {
      this.resolve(this.uploadedParts);
    }
  }
}

export default FileUploader;
