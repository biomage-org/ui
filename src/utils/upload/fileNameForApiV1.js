const fileNameForApiV1 = {
  matrix10x: 'matrix.mtx.gz',
  barcodes10x: 'barcodes.tsv.gz',
  features10x: 'features.tsv.gz',
  seurat: 'r.rds',
  rhapsody: 'expression_data.st.gz',
};

export default fileNameForApiV1;
