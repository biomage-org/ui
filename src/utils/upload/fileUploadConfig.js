const MB = 1024 * 1024;
const fileUploadConfig = {
  chunkSize: 32 * MB,
};

export default fileUploadConfig;
