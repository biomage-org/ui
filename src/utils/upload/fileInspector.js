import { Gunzip } from 'fflate';

import { sampleTech } from 'utils/constants';
import techOptions from 'utils/upload/fileUploadSpecifications';
import readFileToBuffer from 'utils/upload/readFileToBuffer';

const Verdict = {
  INVALID_NAME: -2,
  INVALID_FORMAT: -1,
  VALID_ZIPPED: 0,
  VALID_UNZIPPED: 1,
};

const MATRIX_SIGNATURE = Buffer.from('%%MatrixMarket');

const GZIP_SIGNATURE = Buffer.from([0x1f, 0x8b]);

const inspectFile = async (file, technology) => {
  if (technology === sampleTech['10X']) {
    return inspect10XFile(file, technology);
  } if (technology === sampleTech.SEURAT) {
    return inspectSeuratFile(file, technology);
  } if (technology === sampleTech.RHAPSODY) {
    return inspectRhapsodyFile(file);
  }

  return Verdict.INVALID_FORMAT;
};

const inspect10XFile = async (file, technology) => {
  // Validate a file requested for upload to the platform.

  // immediately discard file if filename is not in valid set
  const validNames = techOptions[technology].acceptedFiles;
  if (!validNames.has(file.name)) {
    return Verdict.INVALID_NAME;
  }

  // if name is valid, inspect first 16 bytes to validate format
  let data = await readFileToBuffer(file.slice(0, 16));

  const isGzipped = !data.slice(0, 2).compare(GZIP_SIGNATURE);

  if (isGzipped) {
    // if gzipped, decompress a small chunk to further validate contents
    const gunzip = new Gunzip((chunk) => {
      data = Buffer.from(chunk.slice(0, 16));
    });
    gunzip.push(await readFileToBuffer(file.slice(0, 128)));
  }

  const valid = isGzipped ? Verdict.VALID_ZIPPED : Verdict.VALID_UNZIPPED;

  // check matrix file starts with matrix signature
  if (file.name.startsWith('matrix')
    && !data.slice(0, MATRIX_SIGNATURE.length).compare(MATRIX_SIGNATURE)) {
    return valid;
  }

  // gene/non-coding IDs can be in many formats so we don't validate features
  if (file.name.startsWith('features') || file.name.startsWith('genes')) {
    return valid;
  }

  // check barcodes file starts with a 16 digit DNA sequence
  if (file.name.startsWith('barcodes')
    && !data.toString().match(/\t/)) {
    return valid;
  }

  return Verdict.INVALID_FORMAT;
};

const inspectSeuratFile = async (file, technology) => {
  // Validate a file requested for upload to the platform.

  // immediately discard file if filename is not in valid set
  const validExtensions = techOptions[technology].validExtensionTypes;

  const isValidExtension = validExtensions.some(
    (validExtension) => file.name.endsWith(validExtension),
  );

  if (!isValidExtension) {
    return Verdict.INVALID_FORMAT;
  }

  return Verdict.VALID_ZIPPED;
};

const inspectRhapsodyFile = async (file) => {
  if (!file.name.toLowerCase().includes('expression_data.st')) {
    return Verdict.INVALID_NAME;
  }

  const data = await readFileToBuffer(file.slice(0, 16));
  const isGzipped = !data.slice(0, 2).compare(GZIP_SIGNATURE);
  const valid = isGzipped ? Verdict.VALID_ZIPPED : Verdict.VALID_UNZIPPED;
  return valid;
};

export {
  inspectFile,
  Verdict,
};
