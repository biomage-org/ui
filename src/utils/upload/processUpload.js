/* eslint-disable no-param-reassign */
import _ from 'lodash';

import {
  createSamples, createSampleFile, updateSampleFileUpload, validateSamples,
} from 'redux/actions/samples';

import UploadStatus from 'utils/upload/UploadStatus';
import loadAndCompressIfNecessary from 'utils/upload/loadAndCompressIfNecessary';
import { inspectFile, Verdict } from 'utils/upload/fileInspector';
import fetchAPI from 'utils/http/fetchAPI';

import getFileTypeV2 from 'utils/getFileTypeV2';
import { sampleTech } from 'utils/constants';
import uploadParts from './processMultipartUpload';

const prepareAndUploadFileToS3 = async (
  projectId, sampleId, fileType, file, uploadUrlParams, dispatch,
) => {
  let parts = null;
  let compressedFile = file;
  const { signedUrls, uploadId, sampleFileId } = uploadUrlParams;

  if (!file.compressed) {
    try {
      compressedFile = await loadAndCompressIfNecessary(file, () => {
        dispatch(updateSampleFileUpload(projectId, sampleId, fileType, UploadStatus.COMPRESSING));
      });
    } catch (e) {
      const fileErrorStatus = e.message === 'aborted' ? UploadStatus.FILE_READ_ABORTED : UploadStatus.FILE_READ_ERROR;

      dispatch(updateSampleFileUpload(projectId, sampleId, fileType, fileErrorStatus));
      return;
    }
  }

  const uploadedPartSizes = new Array(signedUrls.length).fill(0);
  const totalSize = file.fileObject.size;

  const createOnUploadProgressForPart = (partIndex) => (progress) => {
    uploadedPartSizes[partIndex] = progress.loaded;
    const totalUploaded = _.sum(uploadedPartSizes);
    const percentProgress = Math.floor((totalUploaded * 100) / totalSize);

    dispatch(
      updateSampleFileUpload(
        projectId, sampleId, fileType, UploadStatus.UPLOADING, percentProgress ?? 0,
      ),
    );
  };

  try {
    parts = await uploadParts(compressedFile, signedUrls, createOnUploadProgressForPart);
  } catch (e) {
    dispatch(updateSampleFileUpload(projectId, sampleId, fileType, UploadStatus.UPLOAD_ERROR));
    return;
  }

  const requestUrl = '/v2/completeMultipartUpload';
  const body = {
    parts,
    uploadId,
    sampleFileId,
  };

  await fetchAPI(requestUrl,
    {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    });

  dispatch(updateSampleFileUpload(projectId, sampleId, fileType, UploadStatus.UPLOADED));
  return parts;
};

const getMetadata = (file, selectedTech) => {
  const metadata = {};
  if (selectedTech === sampleTech['10X']) {
    if (file.name.includes('genes')) {
      metadata.cellranger_version = 'v2';
    } else if (file.name.includes('features')) {
      metadata.cellranger_version = 'v3';
    }
  }

  return metadata;
};

const createAndUploadSingleFile = async (file, projectId, sampleId, dispatch, selectedTech) => {
  const metadata = getMetadata(file, selectedTech);
  const fileType = getFileTypeV2(file.fileObject.name, file.fileObject.type);

  let uploadUrlParams;
  try {
    uploadUrlParams = await dispatch(
      createSampleFile(
        projectId,
        sampleId,
        fileType,
        file.size,
        metadata,
        file,
      ),
    );
  } catch (e) {
    // If there was an error we can't continue the process, so return
    // (the action creator handles the other consequences of the error)
    return;
  }

  await prepareAndUploadFileToS3(projectId, sampleId, fileType, file, uploadUrlParams, dispatch);
};

const processUpload = async (filesList, technology, samples, experimentId, dispatch) => {
  // First use map to make it easy to add files in the already existing sample entry
  const samplesMap = filesList.reduce((acc, file) => {
    const pathToArray = file.name.trim().replace(/[\s]{2,}/ig, ' ').split('/');

    const sampleName = pathToArray[0];
    const fileName = _.last(pathToArray);

    // Update the file name so that instead of being saved as
    // e.g. WT13/matrix.tsv.gz, we save it as matrix.tsv.gz
    file.name = fileName;

    const sampleUuid = Object.values(samples).filter(
      (s) => s.name === sampleName
        && s.experimentId === experimentId,
    )[0]?.uuid;

    return {
      ...acc,
      [sampleName]: {
        ...acc[sampleName],
        uuid: sampleUuid,
        files: {
          ...acc[sampleName]?.files,
          [fileName]: file,
        },
      },
    };
  }, {});

  const validSamplesList = await dispatch(validateSamples(experimentId, samplesMap, technology));

  // If none of the files are in valid format, return
  if (validSamplesList.length === 0) return;

  // Sort alphabetically
  validSamplesList.sort(([oneName], [otherName]) => oneName.localeCompare(otherName));

  try {
    const sampleIdsByName = await dispatch(
      createSamples(
        experimentId,
        validSamplesList,
        technology,
      ),
    );

    validSamplesList.forEach(([name, sample]) => {
      Object.values(sample.files).forEach((file) => (
        createAndUploadSingleFile(file, experimentId, sampleIdsByName[name], dispatch, technology)
      ));
    });
  } catch (e) {
    // Ignore the error, if createSamples fails we throw to
    // avoid attempting to upload any of these broken samples
  }
};
/**
 * This function converts an uploaded File object into a file record that will be inserted under
 * samples[files] in the redux store.
 * @param {File} fileObject the File object that is uploaded to the .
 * @param {string} technology the chosen technology that outputs this file. Used for verification.
 * @returns {object} fileRecord object that will be associated with a sample.
 */
const fileObjectToFileRecord = async (fileObject, technology) => {
  // This is the first stage in uploading a file.

  // if the file has a path, trim to just the file and its folder.
  // otherwise simply use its name
  const filename = (fileObject.path)
    ? _.takeRight(fileObject.path.split('/'), 2).join('/')
    : fileObject.name;

  const verdict = await inspectFile(fileObject, technology);

  let error = '';
  if (verdict === Verdict.INVALID_NAME) {
    error = 'Invalid file name.';
  } else if (verdict === Verdict.INVALID_FORMAT) {
    error = 'Invalid file format.';
  }

  return {
    name: filename,
    fileObject,
    size: fileObject.size,
    path: fileObject.path,
    upload: {
      status: UploadStatus.UPLOADING,
      progress: 0,
    },
    valid: !error,
    errors: error,
    compressed: verdict === Verdict.VALID_ZIPPED,
  };
};

export {
  fileObjectToFileRecord,
  createAndUploadSingleFile,
};

export default processUpload;
