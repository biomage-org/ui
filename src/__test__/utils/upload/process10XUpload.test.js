import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import waitForActions from 'redux-mock-store-await-actions';
import axios from 'axios';
import fetchMock, { enableFetchMocks } from 'jest-fetch-mock';

import { SAMPLES_FILE_UPDATE } from 'redux/actionTypes/samples';
import initialSampleState, { sampleTemplate } from 'redux/reducers/samples/initialState';
import initialExperimentState, { experimentTemplate } from 'redux/reducers/experiments/initialState';

import UploadStatus from 'utils/upload/UploadStatus';
import { waitFor } from '@testing-library/dom';

import { process10XUpload } from 'utils/upload/processUpload';

import validate from 'utils/upload/sampleValidator';
import pushNotificationMessage from 'utils/pushNotificationMessage';

const FILE_SIZE = 1024;

enableFetchMocks();

const getValidFiles = (cellrangerVersion, compressed = true) => {
  const filename = cellrangerVersion === 'v2' ? 'genes.tsv.gz' : 'features.tsv.gz';

  return ([
    {
      name: `WT13/${filename}`,
      fileObject: mockFile(filename),
      upload: { status: UploadStatus.UPLOADING },
      errors: '',
      compressed,
      valid: true,
    },
    {
      name: 'WT13/barcodes.tsv.gz',
      fileObject: mockFile('barcodes.tsv.gz'),
      upload: { status: UploadStatus.UPLOADING },
      errors: '',
      compressed,
      valid: true,
    },
    {
      name: 'WT13/matrix.mtx.gz',
      fileObject: mockFile('matrix.mtx.gz'),
      upload: { status: UploadStatus.UPLOADING },
      errors: '',
      compressed,
      valid: true,
    },
  ]);
};

const mockFile = (name, path = `/WT13/${name}`, size = FILE_SIZE, mimeType = 'application/gzip') => {
  function range(count) {
    let output = '';
    for (let i = 0; i < count; i += 1) {
      output += 'a';
    }
    return output;
  }

  const blob = new Blob([range(size)], { type: mimeType });
  blob.lastModifiedDate = new Date();
  blob.name = name;
  blob.path = path;

  return blob;
};

const sampleType = '10X Chromium';
const mockSampleUuid = 'sample-uuid';
const mockExperimentId = 'project-uuid';
const sampleName = 'mockSampleName';

const mockUnrelatedSampleUuid = 'unrelated-sample-uuid';
const mockUnrelatedExperimentId = 'unrelated-experiment-id';

const initialState = {
  experiments: {
    ...initialExperimentState,
    [mockExperimentId]: {
      ...experimentTemplate,
      id: mockExperimentId,
    },
  },
  samples: {
    ...initialSampleState,
    ids: [mockSampleUuid],
    meta: {
      loading: true,
      error: false,
    },
    [mockSampleUuid]: {
      ...sampleTemplate,
      uuid: mockSampleUuid,
      name: sampleName,
      experimentId: mockExperimentId,
    },
    [mockUnrelatedSampleUuid]: {
      ...sampleTemplate,
      uuid: mockUnrelatedSampleUuid,
      name: sampleName,
      experimentId: mockUnrelatedExperimentId,
    },
  },
};
// Based on https://stackoverflow.com/a/51045733
const flushPromises = () => new Promise(setImmediate);
const mockStore = configureMockStore([thunk]);

jest.mock('utils/upload/loadAndCompressIfNecessary',
  () => jest.fn().mockImplementation(
    (file) => {
      if (!file.valid) {
        return Promise.reject(new Error('error'));
      }
      return Promise.resolve('loadedGzippedFile');
    },
  ));

jest.mock('uuid', () => ({
  v4: jest.fn(() => 'sample-uuid'),
}));

jest.mock('axios', () => ({
  request: jest.fn(),
}));

jest.mock('utils/pushNotificationMessage');

jest.mock('utils/upload/sampleValidator');

let store = null;

describe('process10XUpload', () => {
  beforeEach(() => {
    jest.clearAllMocks();

    const mockUploadUrlParams = {
      signedUrls: ['theSignedUrl'],
      uploadId: 'some_id',
    };

    fetchMock.resetMocks();
    fetchMock.doMock();
    fetchMock.mockResponse(JSON.stringify(mockUploadUrlParams), { status: 200 });

    store = mockStore(initialState);
  });

  it('Uploads and updates redux correctly when there are no errors with cellranger v3', async () => {
    const mockAxiosCalls = [];
    const uploadSuccess = (params) => {
      mockAxiosCalls.push(params);
      return Promise.resolve({ headers: { etag: 'etag-blah' } });
    };

    axios.request.mockImplementationOnce(uploadSuccess)
      .mockImplementationOnce(uploadSuccess)
      .mockImplementationOnce(uploadSuccess);

    await process10XUpload(
      getValidFiles('v3'),
      sampleType,
      store.getState().samples,
      mockExperimentId,
      store.dispatch,
    );

    // Wait for uploads to be made
    await waitForActions(
      store,
      new Array(3).fill({
        type: SAMPLES_FILE_UPDATE,
        payload: { fileDiff: { upload: { status: UploadStatus.UPLOADED } } },
      }),
      { matcher: waitForActions.matchers.containing },
    );

    // Three axios put calls are made
    expect(mockAxiosCalls.length).toBe(3);
    // Each put call is made with the correct information
    expect(mockAxiosCalls[0].data).toBeInstanceOf(Blob);
    expect(mockAxiosCalls[1].data).toBeInstanceOf(Blob);
    expect(mockAxiosCalls[2].data).toBeInstanceOf(Blob);

    // Wait until all put promises are resolved
    await flushPromises();

    const fileUpdateActions = store.getActions().filter(
      (action) => action.type === SAMPLES_FILE_UPDATE,
    );
    const uploadProperties = fileUpdateActions.map((action) => action.payload.fileDiff.upload);
    const uploadingStatusProperties = uploadProperties.filter(
      ({ status }) => status === UploadStatus.UPLOADING,
    );
    const uploadedStatusProperties = uploadProperties.filter(
      ({ status }) => status === UploadStatus.UPLOADED,
    );

    // There are 3 files actions with status uploading
    expect(uploadingStatusProperties.length).toEqual(3);
    // There are 3 files actions with status uploaded
    expect(uploadedStatusProperties.length).toEqual(3);

    // axios request calls are correct
    expect(axios.request.mock.calls.map((call) => call[0])).toMatchSnapshot();

    // If we trigger axios onUploadProgress it updates the progress correctly
    axios.request.mock.calls[0][0].onUploadProgress({ loaded: FILE_SIZE / 2 });

    await waitForActions(
      store,
      [{
        type: SAMPLES_FILE_UPDATE,
        payload: { fileDiff: { upload: { status: UploadStatus.UPLOADING, progress: 50 } } },
      }],
      { matcher: waitForActions.matchers.containing },
    );
  });

  it('Uploads and updates redux correctly when there are no errors with cellranger v2', async () => {
    const mockAxiosCalls = [];
    const uploadSuccess = (params) => {
      mockAxiosCalls.push(params);
      return Promise.resolve({ headers: { etag: 'etag-blah' } });
    };

    axios.request.mockImplementationOnce(uploadSuccess)
      .mockImplementationOnce(uploadSuccess)
      .mockImplementationOnce(uploadSuccess);

    await process10XUpload(
      getValidFiles('v2'),
      sampleType,
      store.getState().samples,
      mockExperimentId,
      store.dispatch,
    );

    // Wait for uploads to be made
    await waitForActions(
      store,
      new Array(3).fill({
        type: SAMPLES_FILE_UPDATE,
        payload: { fileDiff: { upload: { status: UploadStatus.UPLOADED } } },
      }),
      { matcher: waitForActions.matchers.containing },
    );

    // Three axios put calls are made
    expect(mockAxiosCalls.length).toBe(3);
    // Each put call is made with the correct information
    expect(mockAxiosCalls[0].data).toBeInstanceOf(Blob);
    expect(mockAxiosCalls[1].data).toBeInstanceOf(Blob);
    expect(mockAxiosCalls[2].data).toBeInstanceOf(Blob);

    // Wait until all put promises are resolved
    await flushPromises();

    const fileUpdateActions = store.getActions().filter(
      (action) => action.type === SAMPLES_FILE_UPDATE,
    );
    const uploadProperties = fileUpdateActions.map((action) => action.payload.fileDiff.upload);
    const uploadingStatusProperties = uploadProperties.filter(
      ({ status }) => status === UploadStatus.UPLOADING,
    );
    const uploadedStatusProperties = uploadProperties.filter(
      ({ status }) => status === UploadStatus.UPLOADED,
    );

    // There are 3 files actions with status uploading
    expect(uploadingStatusProperties.length).toEqual(3);
    // There are 3 files actions with status uploaded
    expect(uploadedStatusProperties.length).toEqual(3);

    // axios request calls are correct
    expect(axios.request.mock.calls.map((call) => call[0])).toMatchSnapshot();

    // If we trigger axios onUploadProgress it updates the progress correctly
    axios.request.mock.calls[0][0].onUploadProgress({ loaded: FILE_SIZE / 2 });

    await waitForActions(
      store,
      [{
        type: SAMPLES_FILE_UPDATE,
        payload: { fileDiff: { upload: { status: UploadStatus.UPLOADING, progress: 50 } } },
      }],
      { matcher: waitForActions.matchers.containing },
    );
  });

  it('Updates redux correctly when there are file upload errors', async () => {
    const mockAxiosCalls = [];

    const uploadError = (params) => {
      mockAxiosCalls.push(params);
      return Promise.reject(new Error('Error'));
    };

    axios.request.mockImplementationOnce(uploadError)
      .mockImplementationOnce(uploadError)
      .mockImplementationOnce(uploadError);

    await process10XUpload(
      getValidFiles('v3'),
      sampleType,
      store.getState().samples,
      'errorProjectUuid',
      store.dispatch,
    );

    // Wait for uploads to be made
    await waitForActions(
      store,
      new Array(3).fill({
        type: SAMPLES_FILE_UPDATE,
        payload: { fileDiff: { upload: { status: UploadStatus.UPLOAD_ERROR } } },
      }),
      { matcher: waitForActions.matchers.containing },
    );

    const fileUpdateActions = store.getActions().filter(
      (action) => action.type === SAMPLES_FILE_UPDATE,
    );

    const uploadProperties = fileUpdateActions.map((action) => action.payload.fileDiff.upload);

    const uploadingFileProperties = uploadProperties.filter(
      ({ status }) => status === UploadStatus.UPLOADING,
    );

    const errorFileProperties = uploadProperties.filter(
      ({ status }) => status === UploadStatus.UPLOAD_ERROR,
    );

    const uploadedFileProperties = uploadProperties.filter(
      ({ status }) => status === UploadStatus.UPLOADED,
    );

    // There are 3 files actions with status uploading
    expect(uploadingFileProperties.length).toEqual(3);
    // There are 3 files actions with status upload error
    expect(errorFileProperties.length).toEqual(3);
    // There are no file actions with status successfully uploaded
    expect(uploadedFileProperties.length).toEqual(0);
  });

  it('Should not upload files if there are errors creating samples in the api', async () => {
    fetchMock.mockReject(new Error('Error'));

    await process10XUpload(
      getValidFiles('v3'),
      sampleType,
      store.getState().samples,
      mockExperimentId,
      store.dispatch,
    );

    // We do not expect uploads to happen
    await waitFor(() => {
      expect(axios.request).not.toHaveBeenCalled();
    });
  });

  it('Should not upload sample and show notification if uploaded sample is invalid', async () => {
    validate.mockImplementationOnce(
      () => (['Some file error']),
    );

    await process10XUpload(
      getValidFiles('v2'),
      sampleType,
      store.getState().samples,
      mockExperimentId,
      store.dispatch,
    );

    // We do not expect uploads to happen
    await waitFor(() => {
      expect(pushNotificationMessage).toHaveBeenCalledTimes(1);
      expect(axios.request).not.toHaveBeenCalled();
    });
  });
});
