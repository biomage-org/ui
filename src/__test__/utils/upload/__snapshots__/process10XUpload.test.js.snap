// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`process10XUpload Uploads and updates redux correctly when there are no errors with cellranger v2 1`] = `
Array [
  Object {
    "data": Blob {},
    "headers": Object {
      "Content-Type": "application/octet-stream",
    },
    "method": "put",
    "onUploadProgress": [Function],
    "url": "theSignedUrl",
  },
  Object {
    "data": Blob {},
    "headers": Object {
      "Content-Type": "application/octet-stream",
    },
    "method": "put",
    "onUploadProgress": [Function],
    "url": "theSignedUrl",
  },
  Object {
    "data": Blob {},
    "headers": Object {
      "Content-Type": "application/octet-stream",
    },
    "method": "put",
    "onUploadProgress": [Function],
    "url": "theSignedUrl",
  },
]
`;

exports[`process10XUpload Uploads and updates redux correctly when there are no errors with cellranger v3 1`] = `
Array [
  Object {
    "data": Blob {},
    "headers": Object {
      "Content-Type": "application/octet-stream",
    },
    "method": "put",
    "onUploadProgress": [Function],
    "url": "theSignedUrl",
  },
  Object {
    "data": Blob {},
    "headers": Object {
      "Content-Type": "application/octet-stream",
    },
    "method": "put",
    "onUploadProgress": [Function],
    "url": "theSignedUrl",
  },
  Object {
    "data": Blob {},
    "headers": Object {
      "Content-Type": "application/octet-stream",
    },
    "method": "put",
    "onUploadProgress": [Function],
    "url": "theSignedUrl",
  },
]
`;
